File: app/src/main/java/com/sleticalboy/dailywork/weight/xrefresh/HeaderView.java
package com.sleticalboy.dailywork.weight.xrefresh;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.sleticalboy.dailywork.R;
import com.sleticalboy.dailywork.weight.xrefresh.interfaces.IHeaderView;

/**
 * Created on 18-2-3.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
public class HeaderView extends FrameLayout implements IHeaderView {

    private ImageView default_ptr_flip;
    private ProgressBar default_ptr_rotate;
    private TextView refresh_tip, refresh_time;

    public HeaderView(@NonNull Context context) {
        this(context, null);
    }

    public HeaderView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public HeaderView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        LayoutInflater.from(getContext()).inflate(R.layout.layout_refresh_header, this);
        default_ptr_flip = ((ImageView) findViewById(R.id.default_ptr_flip));
        default_ptr_rotate = ((ProgressBar) findViewById(R.id.default_ptr_rotate));
        refresh_tip = ((TextView) findViewById(R.id.refresh_tip));
        refresh_time = ((TextView) findViewById(R.id.refresh_time));
    }

    @Override
    public void begin() {

    }

    @Override
    public void progress(long progress, long total) {
        if (progress / total >= 0.9f) {
            default_ptr_flip.setRotation(180);
        } else {
            default_ptr_flip.setRotation(0);
        }
        if (progress >= total - 10) {
            refresh_tip.setText("松开以刷新");
        } else {
            refresh_tip.setText("下拉以刷新");
        }
    }

    @Override
    public void finish(long progress, long total) {

    }

    @Override
    public void loading() {
        default_ptr_flip.setVisibility(GONE);
        default_ptr_rotate.setVisibility(VISIBLE);
        refresh_tip.setText("加载中...");
        refresh_time.setText("2018-02-03 20:14:50");
    }

    @Override
    public void hidden() {
        default_ptr_flip.setVisibility(VISIBLE);
        default_ptr_rotate.setVisibility(GONE);
        refresh_tip.setText("下拉以刷新");
        refresh_time.setText("2018-02-03 20:15:03");
    }

    @Override
    public View get() {
        return this;
    }
}

File: MainApp/src/main/java/com/sleticalboy/weight/xrefresh/HeaderView.kt
package com.sleticalboy.weight.xrefresh

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView

import com.sleticalboy.learning.R
import com.sleticalboy.weight.xrefresh.interfaces.IHeaderView

/**
 * Created on 18-2-3.
 *
 * @author leebin
 * @version 1.0
 * @description
 */
class HeaderView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
  defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), IHeaderView {

  private var default_ptr_flip: ImageView? = null
  private var default_ptr_rotate: ProgressBar? = null
  private var refresh_tip: TextView? = null
  private var refresh_time: TextView? = null

  init {
    init()
  }

  private fun init() {
    LayoutInflater.from(context).inflate(R.layout.layout_refresh_header, this)
    default_ptr_flip = findViewById<View>(R.id.default_ptr_flip) as ImageView
    default_ptr_rotate = findViewById<View>(R.id.default_ptr_rotate) as ProgressBar
    refresh_tip = findViewById<View>(R.id.refresh_tip) as TextView
    refresh_time = findViewById<View>(R.id.refresh_time) as TextView
  }

  override fun begin() {

  }

  override fun progress(progress: Long, total: Long) {
    if (progress / total >= 0.9f) {
      default_ptr_flip!!.rotation = 180f
    } else {
      default_ptr_flip!!.rotation = 0f
    }
    if (progress >= total - 10) {
      refresh_tip!!.text = "松开以刷新"
    } else {
      refresh_tip!!.text = "下拉以刷新"
    }
  }

  override fun finish(progress: Long, total: Long) {

  }

  override fun loading() {
    default_ptr_flip!!.visibility = View.GONE
    default_ptr_rotate!!.visibility = View.VISIBLE
    refresh_tip!!.text = "加载中..."
    refresh_time!!.text = "2018-02-03 20:14:50"
  }

  override fun hidden() {
    default_ptr_flip!!.visibility = View.VISIBLE
    default_ptr_rotate!!.visibility = View.GONE
    refresh_tip!!.text = "下拉以刷新"
    refresh_time!!.text = "2018-02-03 20:15:03"
  }

  override fun get(): View {
    return this
  }
}
